SchemaMigration/COSchemaMigrationDriver.m

```diff
@@ -0,0 +1,428 @@
+		[self migrateItemsInDomain: packageName
+						 toVersion: (int64_t)package.version];
+	}
+
+	return [upToDateItems arrayByAddingObjectsFromArray: [self combineMigratedItems: itemsToMigrate]];
+}
+
+#pragma mark Ungrouping Items by Packages -
+
+static inline void copyAttributesFromItemTo(NSArray *attributes, COItem *sourceItem, COMutableItem *destinationItem)
+{
+	for (NSString *attribute in attributes)
+	{
+		[destinationItem setValue: [sourceItem valueForAttribute: attribute]
+		             forAttribute: attribute
+		                     type: [sourceItem typeForAttribute: attribute]];
+	}
+}
+
+static inline COMutableItem *pristineMutableItemFrom(COItem *item)
+{
+	COMutableItem *pristineItem = [COMutableItem itemWithUUID: item.UUID];
+
+	[pristineItem setValue: [item valueForAttribute: kCOObjectEntityNameProperty]
+	          forAttribute: kCOObjectEntityNameProperty
+	                  type: [item typeForAttribute: kCOObjectEntityNameProperty]];
+	[pristineItem setValue: [item valueForAttribute: kCOObjectPackageNameProperty]
+	          forAttribute: kCOObjectPackageNameProperty
+	                  type: [item typeForAttribute: kCOObjectPackageNameProperty]];
+	[pristineItem setValue: [item valueForAttribute: kCOObjectPackageVersionProperty]
+	          forAttribute: kCOObjectPackageVersionProperty
+	                  type: [item typeForAttribute: kCOObjectPackageVersionProperty]];
+
+	return pristineItem;
+}
+
+- (NSArray *)attributesForPackage: (ETPackageDescription *)package inItem: (COItem *)item
+{
+	if ([item isAdditionalItem])
+	{
+		return [item.attributeNames arrayByRemovingObjectsInArray:
+			@[kCOObjectEntityNameProperty, kCOObjectPackageNameProperty, kCOObjectPackageVersionProperty]];
+	}
+
+	ETEntityDescription *entity = [_modelDescriptionRepository descriptionForName: item.entityName];
+	ETAssert(entity != nil);
+	return [entity persistentPropertyDescriptionNamesForPackageDescription: package];
+}
+
+/**
+ * Combines migrated items with the same UUID.
+ *
+ * For each package, items are enumerated, and properties that belong to this
+ * package are merged into a final item per item UUID.
+ */
+- (NSArray *)combineMigratedItems: (NSDictionary *)migratedItems
+{
+	NSMutableDictionary *combinedItems = [NSMutableDictionary new];
+	
+	for (NSString *packageName in migratedItems)
+	{
+		ETPackageDescription *package = [_modelDescriptionRepository descriptionForName: packageName];
+
+		for (COItem *item in migratedItems[packageName])
+		{
+			NSArray *attributes = [self attributesForPackage: package
+			                                          inItem: item];
+			ETAssert(attributes != nil);
+			COMutableItem *combinedItem = combinedItems[item.UUID];
+			
+			if (combinedItem == nil)
+			{
+				combinedItem = pristineMutableItemFrom(item);
+				combinedItems[item.UUID] = combinedItem;
+			}
+			
+			copyAttributesFromItemTo(attributes, item, combinedItem);
+			//[combinedItem setVersion: package.version
```

Comment:
Here I think we miss a bit of code that sets the version correctly based on the final destination version for the package owning the item.
I guess this is not caught by the test suite currently. We probably a multi-steps scenario where one package containing super-entities is incremented lower than the owning one, or some other one akin to this.

I'm a bit sleepy but I think something like below should do the trick:

BOOL isOwningPackage = [packageName isEqualToString: package.name];

if (isOwningPackage)
{
   [combinedItem setVersion: package.version forPackageName: packageName];
}